File: ocaml/disjoint_set.ml
let dsu n =
  object (self: 'self)
    val p = Array.init n (fun i -> i)
    val h = Array.make n 0
    val mutable nc = n
    method findSet x =
      if p.(x) = x then x
      else begin
        let v = self#findSet p.(x) in
        p.(x) <- v;
        v
      end
    method merge x y =
      let i = self#findSet x in
      let j = self#findSet y in
      if i != j then begin
        nc <- nc - 1;
        if h.(i) < h.(j) then p.(i) <- j
        else if h.(i) > h.(j) then p.(j) <- i
        else begin
          p.(i) <- j;
          h.(j) <- h.(j) + 1
        end;
      end
    method components = nc
  end
;;

File: csharp/DisjointSet.cs
class DisjointSet {
  private int[] p;
  private int[] h;
  public DisjointSet(DisjointSet dsu) {
    p = (int[]) dsu.p.Clone();
    h = (int[]) dsu.h.Clone();
  }
  public DisjointSet(int n) {
    this.p = new int[n];
    this.h = new int[n];
    for (int i = 0; i < n; i++) p[i] = i;
  }
  public int FindSet(int x) {
    int y = p[x];
    if (y == x) {
      return x;
    }
    p[x] = FindSet(y);
    return p[x];
  }
  public bool Merge(int x, int y) {
    int i = FindSet(x);
    int j = FindSet(y);
    if (i != j) {
      if (h[i] < h[j]) {
        p[i] = j;
      } else if (h[i] > h[j]) {
        p[j] = i;
      } else {
        p[i] = j;
        h[j]++;
      }
      return true;
    }
    return false;
  }
}
